storage/pc-ata/ata.c

```diff
@@ -0,0 +1,533 @@
+	/* Short delay */
+	usleep(10);
+}
+
+
+static void ata_select(ata_dev_t *dev, uint8_t devnum, uint64_t lba, uint16_t nsectors, uint8_t mode)
+{
+	ata_bus_t *bus = dev->bus;
+	void *base = bus->base;
+	uint16_t c = 0, h = 0, s = 0;
+	uint8_t devsel = DEVSEL_SET0 | DEVSEL_SET1;
+
+	/* Select device number */
+	devsel |= devnum * DEVSEL_DEVNUM;
+
+	/* Select addressing mode */
+	switch (mode) {
+	case AMODE_CHS:
+		/* Convert LBA to CHS */
+		c = (uint16_t)(lba / ((uint64_t)(dev->heads) * (uint64_t)(dev->sectors)));
+		h = (uint16_t)((lba / dev->sectors) % dev->heads) & DEVSEL_HEAD;
+		s = (uint16_t)((lba % dev->sectors) + 1);
+		break;
+
+	case AMODE_LBA28:
+		c = (uint16_t)(lba >> 8);
+		h = ((uint16_t)(lba >> 24) & DEVSEL_HEAD) | DEVSEL_LBA;
+		s = (uint16_t)lba;
+		break;
+
+	case AMODE_LBA48:
+		c = (uint16_t)(lba >> 8);
+		h = DEVSEL_LBA;
+		s = (uint16_t)lba;
+		break;
+	}
+
+	/* Select the device */
+	ata_writereg(base, REG_DEVSEL, h | devsel, 1);
+
+	/* Wait for the device to push its status onto the bus */
+	ata_delay(bus);
+
+	/* Write features */
+	ata_writereg(base, REG_FEATURES, 0, 1);
+
+	/* Write other registers */
+	if (mode == AMODE_LBA48) {
+		ata_writereg(base, REG_NSECTORS,  (nsectors >>  8) & 0xFF, 1);
+		ata_writereg(base, REG_SECTOR,    (lba      >> 24) & 0xFF, 1);
+		ata_writereg(base, REG_LCYLINDER, (lba      >> 32) & 0xFF, 1);
+		ata_writereg(base, REG_HCYLINDER, (lba      >> 40) & 0xFF, 1);
+	}
+	ata_writereg(base, REG_NSECTORS,  nsectors & 0xFF, 1);
+	ata_writereg(base, REG_SECTOR,    (s >> 0) & 0xFF, 1);
+	ata_writereg(base, REG_LCYLINDER, (c >> 0) & 0xFF, 1);
+	ata_writereg(base, REG_HCYLINDER, (c >> 8) & 0xFF, 1);
+}
+
+
+static int ata_wait(ata_bus_t *bus, uint8_t clear, uint8_t set)
+{
+	uint8_t status;
+
+	do {
+		status = (uint8_t)ata_readreg(bus->base, REG_STATUS, 1);
+
+		if (status & STATUS_ERR)
+			return -1;
+
+		if (status & STATUS_DF)
+			return -2;
+
+	} while ((status & clear) || (status & set) != set);
+
+	return 0;
+}
+
+
+static ssize_t ata_pio(ata_dev_t *dev, uint32_t nsectors, uint8_t *buff, uint8_t dir)
+{
+	ata_bus_t *bus = dev->bus;
+	void *base = bus->base;
+	uint32_t i, j;
+	ssize_t err, ret = 0;
+
+	for (i = 0; i < nsectors; i++) {
+		/* Wait until BSY clears and DRQ sets */
+		if ((err = ata_wait(bus, STATUS_BSY, STATUS_DRQ)))
+			return err;
+
+		switch (dir) {
+		case DIR_READ:
+			/* Read sector */
+			for (j = 0; j < dev->secsize / 2; j++)
+				((uint16_t *)(buff + ret))[j] = (uint16_t)ata_readreg(base, REG_DATA, 2);
+			break;
+
+		case DIR_WRITE:
+			/* Write sector */
+			for (j = 0; j < dev->secsize / 2; j++)
+				ata_writereg(base, REG_DATA, ((uint16_t *)(buff + ret))[j], 2);
+			break;
+		}
+
+		ret += dev->secsize;
+	}
+
+	/* Wait until DRQ clears and RDY sets */
+	if ((err = ata_wait(bus, STATUS_DRQ, STATUS_RDY)))
+		return err;
+
+	return ret;
+}
+
+
+static ssize_t ata_access(ata_dev_t *dev, uint64_t lba, uint16_t nsectors, uint8_t cmd, uint8_t *buff)

```

Comment:
We have convention for functions that should be guarded by mutex outside - these should be prefixed with _
static ssize_t _ata_access(...

Reply:
Fixed.